File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {

        ListNode temp = new ListNode(0);
        ListNode current = temp;

        PriorityQueue<ListNode> pq = new PriorityQueue<>((ListNode a, ListNode b) -> a.val - b.val);

        // add all elements in the list
        for (ListNode head : lists){
            if (head != null){
                pq.add(head);
            }
        }

        // Now Deal with values in 
        while (!pq.isEmpty()){
            ListNode smallest = pq.poll();

            current.next = smallest;
            current = current.next; //Moving the current pointer

            if(smallest.next != null){
                pq.add(smallest.next);
            }
        }

        return temp.next;
        
    }
}

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[Optional[ListNode]]
        :rtype: Optional[ListNode]
        """

        heap = []

        for i, l in enumerate(lists):
            if l:
                heapq.heappush(heap,(l.val,i,l))

        temp = ListNode(0)
        current = temp

        while heap:
            val,i,node = heapq.heappop(heap)
            current.next = node
            current = current.next

            if node.next:
                heapq.heappush(heap,(node.next.val, i ,node.next))

        return temp.next
        

        